zerver/views/messages.py

```diff
@@ -625,9 +626,18 @@ def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
                     search_term['operand'] += ' ' + term['operand']
             else:
                 query = builder.add_term(query, term)
+
+            if term['operator'] == 'stream' and len(narrow) == 1:
+                is_stream_narrow = True
+
         if is_search:
             query = builder.add_term(query, search_term)
 
+        if is_stream_narrow:
+            muting_conditions = exclude_muting_conditions(user_profile, narrow)
+            if muting_conditions:
+                query = query.where(*muting_conditions)
+
```

Comment:
I'm curious how the above section will interact with this code:

~~~
 635     if use_first_unread_anchor:
 636         condition = column("flags").op("&")(UserMessage.flags.read.mask) == 0
 637 
 638         # We exclude messages on muted topics when finding the first unread
 639         # message in this narrow
 640         muting_conditions = exclude_muting_conditions(user_profile, narrow)
 641         if muting_conditions:
 642             condition = and_(condition, *muting_conditions)
~~~

I suspect it will just be redundant at times, but we should think this through a bit.

It also seems like `exclude_muting_conditions` may be kind of a broad function to call, when we really only want to exclude topics, and we want to be careful not to exclude the stream here.  We may want to call `exclude_topic_mutes` instead.

@timabbott The other thing that's a bit funny about the code (before these changes) is that we try to apply most conditions to the query in `NarrowBuilder`, and in fact we do that for `exclude_muting_conditions` in `NarrowBuilder.by_in` for the `in:home` case.  Of course, most conditions are composable, which `NarrowBuilder` is good at, whereas topic-muting is a lot more special-case-prone (hence our handling of `first_unread_anchor` in the big view function).

Reply:
Yeah, I think extracting an `exclude_topic_mutes` sub-function would be a good preparatory refactor for this.

Reply:
The function `exclude_topic_mutes` already exists in `zerver/lib/topic_mutes.py`, so it's more about deciding when to call that directly than extracting it.

But, to Tim's broader point, a good way to tackle this kind of problem is to really study the existing code, and sometimes most of the work is adding test coverage and/or extracting functions and/or cleaning up quirks of the existing code.

It's possible we want to try separate these concerns a bit more, all of which sort of happen in `exclude_muting_conditions`:

* converting stream name to id (this should happen upstream)
* deciding whether filtering makes sense (this is done sort of implicitly)
* doing the actual filtering for streams
* doing the actual filtering for topics.

@AasthaGupta This is a pretty tough issue, we may need to think about it some more.

Reply:
@showell What exactly are you suggesting me to do?